homeassistant/components/kostal_plenticore/sensor.py

```diff
@@ -0,0 +1,294 @@
+
+    async def async_added_to_hass(self) -> None:
+        """Register this entity on the Update Coordinator."""
+        self.coordinator.register_entity(self)
+        await super().async_added_to_hass()
+
+    async def async_will_remove_from_hass(self) -> None:
+        """Unregister this entity from the Update Coordinator."""
+        await super().async_will_remove_from_hass()
+        self.coordinator.unregister_entity(self)
+
+    @property
+    def available(self) -> bool:
+        """Return if this entity can be access in the current Plenticore firmware version."""
+        return self._available
+
+    @available.setter
+    def available(self, available) -> None:
+        """Set the availability of this entity."""
+        self._available = available
+
+    @property
+    def scope(self) -> str:
+        """Return the scope of this Sensor Entity."""
+        return SCOPE_PROCESS_DATA
+
+    @property
+    def unique_id(self) -> str:
+        """Return the unique id of this Sensor Entity."""
+        return f"{self.entry_id}_{self._sensor_name}"
+
+    @property
+    def name(self) -> str:
+        """Return the name of this Sensor Entity."""
+        return f"{self.platform_name} {self._sensor_name}"
+
+    @property
+    def unit_of_measurement(self) -> Optional[str]:
+        """Return the unit of this Sensor Entity or None."""
+        return self._sensor_data.get(ATTR_UNIT_OF_MEASUREMENT, None)
+
+    @property
+    def icon(self) -> Optional[str]:
+        """Return the icon name of this Sensor Entity or None."""
+        return self._sensor_data.get(ATTR_ICON, None)
+
+    @property
+    def device_class(self) -> Optional[str]:
+        """Return the class of this device, from component DEVICE_CLASSES."""
+        return self._sensor_data.get(ATTR_DEVICE_CLASS, None)
+
+    @property
+    def entity_registry_enabled_default(self) -> bool:
+        """Return if the entity should be enabled when first added to the entity registry."""
+        return self._sensor_data.get(ATTR_ENABLED_DEFAULT, False)
+
+    @property
+    def state(self) -> Optional[Any]:
+        """Return the state of the sensor."""
+        if self.coordinator.data is None:
+            # None is translated to STATE_UNKNOWN
+            return None
+
+        try:
+            raw_value = self.coordinator.data[self.scope][self.module_id][self.data_id]
+        except KeyError:
+            return STATE_UNAVAILABLE
+
+        return self._formatter(raw_value) if self._formatter else raw_value
+
+    @property
+    def device_info(self) -> Optional[Dict[str, Any]]:
+        """Device info."""
+        return self.coordinator.device_info
+
+
+class PlenticoreSettingSensor(PlenticoreProcessDataSensor):
+    """Representation of a Plenticore setting value Sensor."""
+
+    @property
+    def scope(self) -> str:
+        """Return the scope of this Sensor Entity."""
+        return SCOPE_SETTING
+
+    async def async_set_new_value(self, value) -> None:
```

Comment:
Sensor entities should normally be read only.

Reply:
The entity is actually read-only. The method is called by the service to send a new value to the inverter via the API. Of course, this ultimately leads to the value of the entity also being changed.
I have adopted this pattern from other integrations, e.g. from Sonos. Please give me more information on how this can be implemented correctly.

Reply:
Removed write access via service from this PR for now.